caravel/viz.py

```diff
@@ -994,7 +994,20 @@ def query_obj(self):
     def get_data(self):
         df = self.get_df()
         df.columns = ['source', 'target', 'value']
-        return df.to_dict(orient='records')
+        recs = df.to_dict(orient='records')
+        seen = set()
+        hierarchy = {row['source']: row['target'] for row in recs}
+        def find_loop(source, target):
```

Comment:
might just be me but comments might make this a little more readable. not huge deal.